Dynamic Linux Slaves

jacderida opened the pull request:
Hi Stephen/Calum,

This is for dynamic slaves that are spun up with the EC2 plugin. For now it is limited to Linux. As I discussed with Stephen, the situation for getting it to work with Windows is a bit more complicated and requires a day or 2 dedicated to that, so we agreed to go with a static instance for now.

This time you'll want to kick off a build after you've spun up the infrastructure, so you can see the Linux slaves spin up dynamically. Note that before you start the build you won't see any Linux slaves - that's expected; you'll see a Windows slave that should be connected, and the Mac slave will be marked as offline, as expected for now. The build won't be able to pass the first stage due to the Mac not being available. That's expected. The Windows build will fail because of the missing `mingw` package (it won't be able to find `gcc`).

Some important new instructions have been added to the [README](https://github.com/jacderida/safe-build-infrastructure/tree/dynamic_slaves):

* The issue with the dollar in the Windows password is gone. I just wrapped the Ansible call in a shell script, so that should deal with the dollar OK.
* Before kicking off the infrastructure provisioning from the Bastion, you need to provide the ID of the private subnet for the slaves. The instructions in the README have been updated for this.
* Under 'Configure Jenkins' there is one additional manual step for providing an SSH key for the slaves. It will be possible to automate this, but the key needs to be provided in a very specific format and I didn't want to spend any time on that just now.

When you have the infrastructure up and running, create the pipeline for SCL using the instructions under 'Configure Jenkins', then run it. After a couple of minutes or so you should see 2 Linux slaves appear on the main page in the GUI. At first they will have a red cross indicating Jenkins can't establish an SSH connection, but that's just because it's waiting for the sshd service on the machine to come online. If you don't forget the manual step to provide the private key, they should connect after a little while.

Cheers,

Chris

## Troubleshooting

### Windows Machine Provisioning

This applies to the `make provision-prod-jenkins-environment-aws` part of the process.

A couple of things to look out for:

* There was one occasion this afternoon where I saw the provision for the Windows machine fail sporadically when trying to install the `awscli` package. If it fails to install any packages, just run the provisioning again. The process is idempotent, so you can simply run `make provision-prod-jenkins-environment-aws` to do it again.
* If you see Ansible hang during the 'run get agent secret key script' task, this is most likely because you've provided the wrong hostname for Jenkins (as I did once this morning). Stop the process with `ctrl+c` and then make sure you have the correct IP address for `JENKINS_MASTER_HOSTNAME`.

### Windows SCL Build

If you didn't restart the Windows instance after provisioning, the Windows build steps in the pipeline will fail instantly complaining it can't find Cargo.

### Dynamic Slaves

If they don't show up or end up connecting, you can identify the problem by looking at the Jenkins log. It's likely there was possibly a problem with some of the manual steps.

To see the Jenkins logs:

* SSH to the Bastion host the same way you do for provisioning the infrastructure
* SSH from there to the Jenkins host (get the private IP from the AWS GUI): `ssh ansible@<private ip of jenkins master>`
* Get the ID of the Jenkins container: `sudo docker ps -a`
* Tail the logs: `sudo docker logs -f <jenkins container id>`

If the problem isn't obvious from the log, it'll probably need to wait till I come back.

jacderida committed:
feat/vm: template for docker slave machine

Very simple template that runs the docker-slave playbook against the
machine on Ansible. This should save an AMI for use with the Jenkins
EC2 plugin.

jacderida committed:
fix/vm: wireguard for local machines

The local inventory didn't have a copy of the wireguard group with the
variables. We don't actually need WireGuard for this setup, but it's
better to use a single playbook for everything, so we need a copy of
these.

jacderida committed:
Merge remote-tracking branch 'upstream/master' into dynamic_slaves

jacderida committed:
fix/vm: replace all references to 7.5 to 7.6

I've actually always been using 7.6, I just thought it was 7.5 at the
time.

jacderida committed:
fix/vm: these should not have been changed

jacderida committed:
feat/vm: install ec2 plugin for jenkins

There's also another little commit here to change the plugin helper
script to allow a remote instance of Jenkins to be used.

jacderida committed:
feat/role: add jenkins ec2 key pair to user

The EC2 plugin for Jenkins appears to only let you use an AWS keypair
and not a set of Jenkins credentials, so the public part of this keypair
needs to be added to the jenkins user.

jacderida committed:
feat/env: create security group for linux slaves

Adds a security group for Linux slaves to the dev environment.

Just to keep things consistent, it also renames the slaves to 001
and 002 as per the production environment.

jacderida committed:
feat/env: slave management with ec2 plugin

This configures the Jenkins EC2 plugin for dynamically spinning up
slaves. There is a new `jenkins-ec2` user created just for this purpose,
and it's locked down to only the operations it needs for this.

This removes the need for the static slaves. So far it's working for
Linux on the default-VPC. The next step is to change to the production
VPC and launch on the private subnet.

jacderida committed:
fix/env: provide private key manually

Unfortunately the plugin seems to require the text content of this key
to be presented as a specific string with newlines in it in the right
places, and the jinja template variable replacement isn't quite putting
it in the correct format. We can just do this manually for now.

jacderida committed:
fix/env: enable docker on startup

This ensure Docker is running when the machine is launched.

jacderida committed:
fix/env: correct content and path

jacderida committed:
fix/env: update dev ami reference

jacderida committed:
feat/env: util target for bringing up master

This is for testing the master individually.

jacderida committed:
Merge remote-tracking branch 'upstream/master' into dynamic_slaves

jacderida committed:
refactor/jenkins: use static windows slave

We have agreed that it will take some more time to figure out how to get
the EC2 plugin to work with Windows. You can't really use a static slave
agent service because it has a reference to a Jenkins master, which
would need to be injected in every time the machine is spun up. I think
the EC2 plugin does its own agent management.

This also introduces a wrapper script for provisioning the Windows
production instance, to get round the problem with the dollar in the
password.

jacderida committed:
fix/script: use correct playbook reference

jacderida committed:
feat/env: use dynamic linux slaves

This removes the creation of Linux slaves as part of the infrastructure
provision, since they'll be spun up dynamically.

There are also some additional instructions for providing the ID of the
subnet to the provisioning process.

jacderida committed:
fix/vm: remove mingw package

This has decided to randomly stop working today, complaining it can't
retrieve the package from the URL. Seems to work fine when you run it
manually. I tried to replace it with the msys2 package but that seemed
to be taking over 15 minutes to install, so I'm not sure if it's started
prompting for user input or something.

We can add this back in later. It will stop SCL from building
successfully, but it won't stop the slave coming up.

jacderida committed:
fix/env: remove provisioning docker slaves

These will now be provided dynamically by the EC2 plugin.

jacderida committed:
refactor/env: rename security groups

Append the environment name to the security group, so that it can
co-exist alongside a development environment.

jacderida committed:
fix/env: correct name reference

jacderida committed:
docs/env: instructions for ec2 plugin

jacderida committed:
docs/env: remove password warning

This no longer applies after wrapping the Ansible call in a script.

calumcraig commented on README.md:
```diff
@@ -119,6 +117,9 @@ Now perform the following steps (all of these have to be applied to the Bastion
 * Get the Windows password from the AWS GUI and set that: `export JENKINS_WINDOWS_SLAVE_PASSWORD='<windows password>'` (note the password must go inside single quotes to prevent Bash from interpreting special characters)
+* Get the ID of the private subnet for the slaves and set that:
+    - In the AWS GUI go to Services -> VPC -> Subnets -> jenkins_environment-private-eu-west-2 and copy the ID of the subnet
+    - `export SLAVE_SUBNET_ID='<private subnet ID>'`
```
Need to remove the single quotes around `<private subnet ID>`

S-Coyle replied:
Note that on macOS I tried this with the single quotes (') included around the subnet ID and it worked fine

jacderida replied:
I'll remove this just for consistency anyway, since it was called out for the password because that's strictly non-optional.

calumcraig reviewed (changes requested):
I've commented on line 122 of README

calumcraig reviewed (commented):
Have successfully made the build environment and provisioned Jenkins as described in this PR. After kicking off an SCL build job in Jenkins, two Linux slaves did spin up dynamically and come online after some time, resulting in the final configuration:
![image](https://user-images.githubusercontent.com/44327924/55395151-e7fa9780-5538-11e9-9106-ca9347f7c2c0.png)

calumcraig commented:
Upon trying to tear down this set up using `cd terraform/prod && terraform destroy` I had some errors:
```2 error(s) occurred:

* aws_security_group.linux_slaves (destroy): 1 error(s) occurred:

* aws_security_group.linux_slaves: DependencyViolation: resource sg-0b0e8e8ee2b54f843 has a dependent object
	status code: 400, request id: e993c1e7-3265-459e-a3dc-761659cbc8f0
* module.vpc.aws_subnet.private (destroy): 1 error(s) occurred:

* aws_subnet.private: Error deleting subnet: timeout while waiting for state to become 'destroyed' (last state: 'pending', timeout: 10m0s)```
And the two slaves were still seen as running in the AWS GUI. I was able to terminate them from the GUI however.

S-Coyle reviewed (dismissed):
✅ Was able to follow the instructions and see the Linux slaves spin up when I kicked off a SCL job.
✅ Cancelled the running jobs & waited 30 minutes to confirm that the Linux slaves were torn down automatically.
✅ Once both Linux slaves were torn down I was able to run `terraform destroy` and successfully remove the remaining infrastructure.

👍

S-Coyle committed:
fix/vagrantfile: fix synced folders issue affecting macOS

S-Coyle reviewed (dismissed):
Also now tested synced folders fix on macOS - works as expected.

jacderida committed:
docs/fix: remove quotes for consistency

The quotes were added because this is strictly necessary for the
password, but it's not necessary for the subnet, and all the other
variables that don't require the quotations don't say you should add
them.

jacderida committed:
Merge branch 'dynamic_slaves' of github.com:jacderida/safe-build-infrastructure into dynamic_slaves

S-Coyle reviewed (approved):
Approved 👍 
Note that I've not tried exporting the subnet ID without the single quotation marks, as per the most recent commits added, but can't see this being a problem as there shouldn't be special characters in this variable
